great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py

```diff
@@ -0,0 +1,247 @@
+from typing import Dict, List, Optional, Union
+
+from great_expectations.core.batch import BatchRequest, RuntimeBatchRequest
+from great_expectations.exceptions import ProfilerConfigurationError
+from great_expectations.execution_engine.execution_engine import MetricDomainTypes
+from great_expectations.rule_based_profiler.domain_builder import DomainBuilder
+from great_expectations.rule_based_profiler.domain_builder.cardinality_checker import (
+    CardinalityChecker,
+    CardinalityLimitMode,
+)
+from great_expectations.rule_based_profiler.domain_builder.domain_builder import (
+    build_domains_from_column_names,
+)
+from great_expectations.rule_based_profiler.types import Domain, ParameterContainer
+from great_expectations.validator.metric_configuration import MetricConfiguration
+from great_expectations.validator.validator import Validator
+
+
+class CategoricalColumnDomainBuilder(DomainBuilder):
+    """
+    This DomainBuilder uses column cardinality to identify domains.
+    """
+
+    def __init__(
+        self,
+        data_context: "DataContext",  # noqa: F821
+        batch_request: Optional[Union[BatchRequest, RuntimeBatchRequest, dict]] = None,
+        limit_mode: Optional[Union[CardinalityLimitMode, str]] = None,
+        max_unique_values: Optional[int] = None,
+        max_proportion_unique: Optional[int] = None,
+        exclude_columns: Optional[List[str]] = None,
+    ):
+        """Create column domains where cardinality is within the specified limit.
+
+        Cardinality refers to the number of unique values in a given domain.
+        Categorical generally refers to columns with relatively limited
+        number of unique values.
+        Limit mode can be absolute (number of unique values) or relative
+        (proportion of unique values). You can choose one of: limit_mode,
+        max_unique_values or max_proportion_unique to specify the cardinality
+        limit.
+        Note that the limit must be met for each batch separately that is
+        supplied in the batch_request or the column domain will not be included.
+        Note that the columns used will be from the first batch retrieved
+        via the batch_request. If other batches contain additional columns,
```

Comment:
@anthonyburdi I am not seeing `batch` and `batch_request` property getters/setters -- I think that we need them for Self-Initializing Expectations (please see `ColumnDomainBuilder`).  Thanks!

Reply:
Could we split that into a Self-initializing expectations PR? I want to understand it better before adding.

Reply:
@anthonyburdi Then we would need to merge #4282 and #4258 prior to this PR -- which I would be grateful for if you can hang on.  Thanks!

Reply:
@anthonyburdi and @alexsherstinsky discussed and agreed to add capabilities needed to enable Self-Initializing expectations that use `ValueSet` in a separate work item.